controllers/k8ssandracluster_controller.go

```diff
@@ -169,113 +172,118 @@ func (r *K8ssandraClusterReconciler) reconcile(ctx context.Context, kc *api.K8ss
 					return ctrl.Result{}, err
 				}
 				return ctrl.Result{RequeueAfter: defaultDelay}, nil
 			} else {
-				logger.Error(err, "Failed to get datacenter", "CassandraDatacenter", dcKey)
+				logger.Error(err, "Failed to get datacenter")
 				return ctrl.Result{}, err
 			}
 		}
 
-		stargateTemplate := dcTemplate.Stargate.Coalesce(kc.Spec.Stargate)
-
-		stargateKey := types.NamespacedName{
-			Namespace: actualDc.Namespace,
-			Name:      kc.Name + "-" + actualDc.Name + "-stargate",
+		result, err := r.reconcileStargate(ctx, kc, dcTemplate, actualDc, logger, remoteClient)
+		if !result.IsZero() || err != nil {
+			return result, err
 		}
-		actualStargate := &api.Stargate{}
-
-		if stargateTemplate != nil {
-
-			desiredStargate := &api.Stargate{
-				ObjectMeta: metav1.ObjectMeta{
-					Namespace:   stargateKey.Namespace,
-					Name:        stargateKey.Name,
-					Annotations: map[string]string{},
-					Labels: map[string]string{
-						api.NameLabel:             api.NameLabelValue,
-						api.PartOfLabel:           api.PartOfLabelValue,
-						api.ComponentLabel:        api.ComponentLabelValueStargate,
-						api.CreatedByLabel:        api.CreatedByLabelValueK8ssandraClusterController,
-						api.K8ssandraClusterLabel: kcKey.Name,
-					},
-				},
-				Spec: api.StargateSpec{
-					StargateDatacenterTemplate: *stargateTemplate,
-					DatacenterRef:              corev1.LocalObjectReference{Name: actualDc.Name},
-				},
-			}
-			desiredStargateHash := utils.DeepHashString(desiredStargate)
-			desiredStargate.Annotations[api.ResourceHashAnnotation] = desiredStargateHash
-
-			if err := remoteClient.Get(ctx, stargateKey, actualStargate); err != nil {
-				if errors.IsNotFound(err) {
-					logger.Info("Creating Stargate resource", "Stargate", stargateKey)
-
-					if err := remoteClient.Create(ctx, desiredStargate); err != nil {
-						logger.Error(err, "Failed to create Stargate resource", "Stargate", stargateKey)
-						return ctrl.Result{}, err
-					} else {
-						return ctrl.Result{RequeueAfter: defaultDelay}, nil
-					}
-				} else {
-					logger.Error(err, "Failed to get Stargate resource", "Stargate", stargateKey)
-					return ctrl.Result{}, err
-				}
+	}
+	kcLogger.Info("Finished reconciling the k8ssandracluster")
+	return ctrl.Result{}, nil
+}
+
+func (r *K8ssandraClusterReconciler) reconcileStargate(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	dcTemplate api.CassandraDatacenterTemplateSpec,
+	actualDc *cassdcapi.CassandraDatacenter,
+	logger logr.Logger,
+	remoteClient client.Client,
+) (ctrl.Result, error) {
+
+	stargateTemplate := dcTemplate.Stargate.Coalesce(kc.Spec.Stargate)
+	stargateKey := types.NamespacedName{
+		Namespace: actualDc.Namespace,
+		Name:      kc.Name + "-" + actualDc.Name + "-stargate",
+	}
+	actualStargate := &api.Stargate{}
+	logger = logger.WithValues("Stargate", stargateKey)
+
+	if stargateTemplate != nil {
+
+		if !kc.Status.IsStargateAuthKeyspaceCreated() {
```

Comment:
More and more I am uncertain about looking at the status of an object to determine whether a change should be applied. We have seen a similar approach in medusa-operator cause problems. 

Here is my reasoning. The status describes the current status of the object. When a Reconciler is called it fetches the object to be reconciled. The status of the object at that point might be stale. There was some event that triggered the reconciliation and now as part of the reconciliation we may recompute and change the status.

Let's look at the situation in practice. Suppose after the auth keyspace is created I accidentally drop it. Will the controller recreate it? I think it should. If we rely on the status condition check, then then controller won't recreate it.

Reply:
I agree with your reasoning. But how can we check the existence of a keyspace inside the reconcile loop? The management API does not offer such an endpoint. I could use the go-cql driver but I would rather avoid it. I could also use go-cassandra-native-protocol [simple client](https://github.com/datastax/go-cassandra-native-protocol/tree/main/client), but it has been designed for testing purposes. Do you have other options in mind?

Reply:
Unfortunately I do not have any other options in mind. I suppose one option could be to always do a `CREATE IF NOT EXISTS`.

Reply:
Unfortunately the management API does not issue the `CREATE KEYSPACE` statement with the `IF NOT EXISTS` option, see [here](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/0608d0c32e8f7d1bb51e5d9b62e05eb767474a4f/management-api-agent-common/src/main/java/com/datastax/mgmtapi/NodeOpsProvider.java#L354-L357).

Reply:
Oh sorry my bad: it does use a `IF NOT EXISTS` option, the right code to look at is [here](https://github.com/k8ssandra/management-api-for-apache-cassandra/blob/0608d0c32e8f7d1bb51e5d9b62e05eb767474a4f/management-api-agent-common/src/main/java/com/datastax/mgmtapi/NodeOpsProvider.java#L336-L340). 

So that might be a quick win.

Reply:
I suggest the following course of action:

1. As a temporary workaround, I will simply remove the condition I created and call the `CreateKeyspace` endpoint for _every_ Stargate resource that the controller creates. 
2. As a long-term solution, we are going to create a new Management API endpoint to list existing keyspaces, then call the `CreateKeyspace` endpoint only if the Stargate auth keyspace is not present yet.

As I'm stepping out for a few days, @adejanovski offered to resume the work where I left. I think though that we could merge this PR with the temporary workaround, then Alex would open a follow-up PR when the endpoint is available. How does that sound @jsanda @adejanovski ?

Reply:
Sounds good to me! Can you create an issue for the follow up work.

Reply:
There you go: https://github.com/k8ssandra/k8ssandra-operator/issues/106.